File: backend/pkg/booking/booking.go
package booking

import "time"

type TimeFrame struct {
	From time.Time
	To   time.Time
}

type Booking interface {
	GetTimeFrame() TimeFrame

	GetBookable() Bookable

	SetBookable(bookable Bookable)

	GetBooker() Booker

	SetBooker(booker Booker)
}

type Bookable interface {
	GetBookings() []Booking

	GetDuration() time.Duration
}

type Booker interface {
	Book(bookable Bookable, timeframe TimeFrame) error

	GetBookings() []Booking
}

File: v0-php/jayplus-dev/packages/booking/src/Eloquent/Booking.php
<?php

namespace Booking\Eloquent;

use Booking\Contracts\Bookable as BookableContract;
use Booking\Contracts\Booker;
use Booking\Contracts\Booking as BookingContract;
use Booking\TimeFrame;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\MorphTo;

class Booking extends Model implements BookingContract
{
    use HasFactory;

    public function getTimeFrame(): TimeFrame
    {
        return new TimeFrame($this->starts_at, $this->ends_at);
    }

    public function getBookable(): BookableContract
    {
        return $this->bookable;
    }

    public function getBooker(): Booker
    {
        return $this->booker;
    }

    /**
     * Get the owning resource model.
     *
     * @return MorphTo
     */
    public function bookable(): MorphTo
    {
        return $this->morphTo();
    }

    /**
     * Get the booker.
     *
     * @return \Illuminate\Database\Eloquent\Relations\MorphTo
     */
    public function booker(): MorphTo
    {
        return $this->morphTo();
    }
}
